README.md
# Block-Tower-Defense
A Tower Defense game in the Alpha stages (V 0.8a1)

Currently: 9 Towers each with 8 upgrades, 10 Different monsters to defend the world from, and 25 Pre-programmed waves with an endless freeplay

Just download the files and run main.py (Pygame required)

V 0.8a1: Second Map, 1 Save File for each map, 3 Different Difficulties, New Cannon Tower, Many Many Bug Fixes, and tons more released on this update!


Changelog
V 0.1a1:
Base game was released, including:
6 Monsters released
2 Towers added
Preprogrammed waves added

V 0.2a1:
Pyro Update:
Flamethrower was added

Bug Fixes:
Fixed infite-casting abilities

V 0.2a2:
Bug Fixes:
Fixed bug which made Ninja and Base tower not allowed to hit monsters

V 0.3a1:
Frozen Metal Update:
Flamethrower speed doubled
Flamethrower range upgrade cost 250 -> 100
Flamethrower speed upgrade cost 100 -> 250
Swapped Flamethrower left 1st and 2nd upgrades due to range being very important on Flamethrower
Fixed fire bugs concerning black monsters popping into two
Added Ice Tower and Lead Monster

V 0.4a1:
Exploding Money Update:
Added Money tower
Added Explosiong Factor
Fixed bugs with infinite abilities

v 0.5a1:
Super Glaive Update:
Super Tower added
Glaive Tower added


v 0.6a1:
Tanky Purple Update:
Purple Monster added
Tank Monster added

v 0.6a2:
Split main.py into 3 files for easier viewing
Squashed some bugs

v 0.6a3:
Added another file for the generation of waves
Added a button to open up settings 
Made all the buttons more clear on what they did
Added flamethrower sprite, the first of many towers to have their graphics updated.

V 0.6a4:
Added dollar and lives symbol
Added in a base Ice sprite(but It'll probably be changed later)
Added ninja tower sprite

V 0.7a1:
Added main menu!
Also added the ability to save your game which happens automatically after every round

V 0.7a2:
Finally taught the clumsy towers how to aim!
Buffed ice tower for early game
Squashed some bugs for the super towers infite range

V 0.8a1:
Coming with this release is a second map!
You now have the ability to keep multiple saves, one for each map!
added difficulties, easy, medium, and hard modes to challenge yourself!
involved in the new update, is the cannon tower! Throwing cannonballs which pop many monsters around it
squashed many of those pesky bugs


waves.py
import monsters, random
def genEnemies(wave, Images, mapName):
    """Outputs a list of all enemies based on a certain wave"""
    Monsters = []
    if wave == 1:
        for i in range(20):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*i])
    elif wave == 2:
        for i in range(35):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),15*i])
    elif wave == 3:
        for i in range(5):
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),60*i])
        for i in range(20):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(i+21)])
    elif wave == 4:
        for j in range(5):
            for i in range(2):
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),45*(i+j*10)])
            for i in range(8):
                Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(i+j*10+2)])
    elif wave == 5:
        for j in range(3):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(1+j*10)])
            for i in range(9):
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i+1+j*10)])
    elif wave == 6:
        for j in range(3):
            for i in range(5):
                Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(i*2+j*11)])
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i*2+1+j*11)])
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(10+j*11)])
        Monsters.append([monsters.Monster(3, wave, False, Images, mapName),60*(18)])
    elif wave == 7:
        for j in range(5):
            for i in range(4):
                Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(i*2+j*9)])
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i*2+1+j*9)])
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(8+j*9)])
    elif wave == 8:
        for i in range(10):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*i])
        for i in range(20):
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i+10)])
        for i in range(14):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i+30)])
    elif wave == 9:
        for i in range(30):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),45*i])
    elif wave == 10:
        for i in range(102):
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),20*i])
    elif wave == 11:
        for j in range(3):
            for i in range(2):
                Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(i+j*11)])
            for i in range(4):
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i*2+j*11)])
                Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i*2+1+j*11)])
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(10+j*11)])
    elif wave == 12:
        for j in range(5):
            for i in range(3):
                Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i+j*7)])
            for i in range(2):
                Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i+3+j*7)])
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(6+j*7)])
    elif wave == 13:
        for i in range(50):
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),15*(i)])
        for i in range(23):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),15*(i+50)])
    elif wave == 14:
        for i in range(10):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i)])
        for i in range(20):
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(i+10)])
        for i in range(8):
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(i+30)])
    elif wave == 15:
        for j in range(5):
            Monsters.append([monsters.Monster(1, wave, False, Images, mapName),30*(1+j*5)])
            Monsters.append([monsters.Monster(2, wave, False, Images, mapName),30*(2+j*5)])
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(3+j*5)])
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(4+j*5)])
            Monsters.append([monsters.Monster(5, wave, False, Images, mapName),30*(5+j*5)])
    elif wave == 16:
        for j in range(8):
            for i in range(5):
                Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i+j*6)])
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(5+j*6)])
    elif wave == 17:
        for i in range(12):
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(i)])
    elif wave == 18:
        for i in range(80):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),20*(i)])
    elif wave == 19:
        for i in range(5):
            Monsters.append([monsters.Monster(3, wave, False, Images, mapName),30*(i)])
        for i in range(7):
            Monsters.append([monsters.Monster(5, wave, False, Images, mapName),30*(i+5)])
        for i in range(10):
            Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(i+12)])
    elif wave == 20:
        for i in range(5):
            Monsters.append([monsters.Monster(6, wave, False, Images, mapName),30*(i)])

        Monsters.append([monsters.Monster(7, wave, False, Images, mapName), 150])
    elif wave == 21:
        for j in range(7):
            for i in range(5):
                Monsters.append([monsters.Monster(4, wave, False, Images, mapName),30*(i+j*6)])
            Monsters.append([monsters.Monster(5, wave, False, Images, mapName),30*(5+j*6)])
    elif wave == 22:
        for i in range(16):
            Monsters.append([monsters.Monster(6, wave, False, Images, mapName),60*(i)])
    elif wave == 23:
        for i in range(7):
            Monsters.append([monsters.Monster(5, wave, False, Images, mapName),30*(i*2)])
            Monsters.append([monsters.Monster(6, wave, False, Images, mapName),30*(i*2+1)])
    elif wave == 24:
        for i in range(5):
            Monsters.append([monsters.Monster(3, wave, True, Images, mapName),30*(i)])
    elif wave % 25 == 0 and wave != 25:
        Monsters.append([monsters.Monster(10, wave, False, Images, mapName),30])
    else:
        for i in range(random.randint(5,15)):
            n = random.randint(1, 100)
            if n <= 40-wave:
                Monsters.append([monsters.Monster(1, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 50-wave:
                Monsters.append([monsters.Monster(2, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 65-wave:
                Monsters.append([monsters.Monster(3, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 75-wave:
                Monsters.append([monsters.Monster(4, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 85-wave:
                Monsters.append([monsters.Monster(5, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 100-wave:
                Monsters.append([monsters.Monster(6, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 125-wave:
                Monsters.append([monsters.Monster(7, wave, False, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 130-wave:
                Monsters.append([monsters.Monster(8, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 175-wave: 
                Monsters.append([monsters.Monster(9, wave, random.randint(1,10)==1, Images, mapName),random.randint(15,30)*(i)])
            elif n <= 200-wave:
                Monsters.append([monsters.Monster(10, wave, False, Images, mapName),random.randint(15,30)*i])
    return Monsters


monster.py
import pygame, random, sys, math, os
from pygame.locals import *
 
pygame.init()


class Monster():
    """Class for all the enemies"""
    def __init__(self, rank, wave, camo, Images):

        #Setting up base variables
        self.rank = rank
        self.x, self.y = -100, 200
        self.step = 0
        speed = [2, 3, 4, 5, 6, 4, 3, 1, 2, 3]
        self.speed = speed[self.rank-1] * (1 + (0.01*(int(wave/50)+1)*wave))
        self.wave = wave
        self.height, self.width = 30, 30
        self.dead = False
        Colors = [(200, 0, 0), (0, 0, 200), (0, 200, 0), (255, 255, 0), (255,105,180), (0, 0, 0), (50, 50, 50), (128, 0, 128), (0, 100, 0), (150, 150, 150)]
        self.color = Colors[self.rank-1]
        self.camo = camo
        self.hit = []
        self.fire = [False,0,0,0,0]
        self.speedModifier = [1,0]
        self.permaslow = False
        self.confused = False
        self.health = 1
        if self.rank == 9:
            self.health = 3
        elif self.rank == 10:
            self.health = 50
        self.addMonster = []
        #All checkpoints on the map
        self.checkpoints = [(6, 5), (6, 2), (3, 2), (3, 9), (13, 9), (13, 4), (17, 4)]
        self.duplicate = False
        self.cooldown = 0
        self.img = Images

        #Setting up the hitbox
        self.image = pygame.Surface([self.width, self.height])
        self.rect = self.image.get_rect()
        self.rect.top = self.y
        self.rect.bottom = self.y + self.height
        self.rect.left = self.x
        self.rect.right = self.x + self.width

    def ReRank(self, new_Rank):
        """Changes the rank of the monster to any other rank"""


        if self.rank == 8 and new_Rank != 8:
            self.addMonster.append(5)
        elif self.rank == 10 and new_Rank != 10:
            for i in range(4):
                self.addMonster.append(9)
        if new_Rank <= 0:
            self.dead = True
        else:
            self.rank = new_Rank
            Colors = [(200, 0, 0), (0, 0, 200), (0, 200, 0), (255, 255, 0), (255,105,180), (0, 0, 0), (50, 50, 50), (128, 0, 128), (0, 100, 0), (150, 150, 150)]
            self.color = Colors[self.rank-1]
            speed = [2, 3, 4, 5, 6, 4, 3, 1, 2, 3]
            self.speed = speed[self.rank-1] * (1 + (0.01*(int(self.wave/50)+1)*self.wave))
            if self.rank == 9:
                self.health = 3
            elif self.rank == 10:
                self.health = 50
        
        
    def draw(self, gameDisplay):

 
        if self.rank != 10:    
            pygame.draw.rect(gameDisplay, self.color, (self.x + int(self.width/5), self.y + int(self.height/5), self.width, self.height), 0)
        else:
            pygame.draw.rect(gameDisplay, self.color, (self.x - int(self.width/3), self.y - int(self.height/3), self.width*2, self.height*2), 0)

        #Drawing camo bloons
        if self.camo:
            if self.rank != 6:
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.8),int(self.color[1]*0.8),int(self.color[2]*0.8)), (self.x + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.5),int(self.color[1]*0.5),int(self.color[2]*0.5)), (self.x + int(self.width/3) + int(self.width/5), self.y + int(self.height/3) + int(self.height/5), int(self.width/3), int(self.height/3)), 0)    
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.75),int(self.color[1]*0.75),int(self.color[2]*0.5)), (self.x + int(self.width/5), self.y + int(self.height/3) + int(self.height/5), int(self.width/3), int(self.height/3)), 0) 
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.5),int(self.color[1]*0.5),int(self.color[2]*0.5)), (self.x + int(self.width/3) + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.3),int(self.color[1]*0.3),int(self.color[2]*0.3)), (self.x + int(self.width/3)*2 + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.5),int(self.color[1]*0.5),int(self.color[2]*0.5)), (self.x + int(self.width/3)*2 + int(self.width/5), self.y + int(self.height/5)+int(self.width/3)*2, int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]*0.75),int(self.color[1]*0.8),int(self.color[2]*0.75)), (self.x + int(self.width/5), self.y + int(self.height/5)+int(self.width/3)*2, int(self.width/3), int(self.height/3)), 0) 
            else:
                pygame.draw.rect(gameDisplay, (int(self.color[0]+50),int(self.color[1]+50),int(self.color[2]+50)), (self.x + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]+30),int(self.color[1]+30),int(self.color[2]+30)), (self.x + int(self.width/3) + int(self.width/5), self.y + int(self.height/3) + int(self.height/5), int(self.width/3), int(self.height/3)), 0)    
                pygame.draw.rect(gameDisplay, (int(self.color[0]+50),int(self.color[1]+50),int(self.color[2]+50)), (self.x + int(self.width/5), self.y + int(self.height/3) + int(self.height/5), int(self.width/3), int(self.height/3)), 0) 
                pygame.draw.rect(gameDisplay, (int(self.color[0]+50),int(self.color[1]+50),int(self.color[2]+50)), (self.x + int(self.width/3) + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]+70),int(self.color[1]+70),int(self.color[2]+70)), (self.x + int(self.width/3)*2 + int(self.width/5), self.y + int(self.height/5), int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]+50),int(self.color[1]+50),int(self.color[2]+50)), (self.x + int(self.width/3)*2 + int(self.width/5), self.y + int(self.height/5)+int(self.width/3)*2, int(self.width/3), int(self.height/3)), 0)
                pygame.draw.rect(gameDisplay, (int(self.color[0]+50),int(self.color[1]+50),int(self.color[2]+50)), (self.x + int(self.width/5), self.y + int(self.height/5)+int(self.width/3)*2, int(self.width/3), int(self.height/3)), 0) 



        if self.fire[0]:
            colors = [(255,0,0), (255,165,0), (255,255,0)]
            for i in range(9):
                pygame.draw.rect(gameDisplay, colors[random.randint(0,2)], (self.x+random.randint(5, 20), self.y+random.randint(5, 20), random.randint(5, 20), random.randint(5, 20)), 0)

        if self.rank != 10:
            pygame.draw.rect(gameDisplay, (100, 100, 100), (self.x + int(self.width/5), self.y + int(self.height/5), self.width, self.height), 2)
        else:
    
            pygame.draw.rect(gameDisplay, (100, 100, 100), (self.x - int(self.width/3), self.y - int(self.height/3), self.width*2, self.height*2), 2)

    def movement(self, Lives, speed):

        #Checking to see whether the monster is in range of the next checkpoint
        if not self.confused:
            if (self.checkpoints[self.step][0]*40-self.speed <= self.x <= self.checkpoints[self.step][0]*40+self.speed and
                self.checkpoints[self.step][1]*40-self.speed <= self.y <= self.checkpoints[self.step][1]*40+self.speed):
                self.x, self.y = self.checkpoints[self.step][0]*40, self.checkpoints[self.step][1]*40
                self.step += 1
                if self.step == 7:
                    modifier = 1
                    if self.rank == 10:
                        modifier = 4
                    self.dead = True
                    Lives -= self.rank*modifier
            else:
                #Else move it towards it's next checkpoint
                if self.x < self.checkpoints[self.step][0]*40:
                    self.x += int(self.speed*speed*self.speedModifier[0])
                elif self.x > self.checkpoints[self.step][0]*40:
                    self.x -= int(self.speed*speed*self.speedModifier[0])
                elif self.y > self.checkpoints[self.step][1]*40:
                    self.y -= int(self.speed*speed*self.speedModifier[0])
                elif self.y < self.checkpoints[self.step][1]*40:
                    self.y += int(self.speed*speed*self.speedModifier[0])
        else:
            if self.step == 0:
                xCheck, yCheck = (-10000, 0)
            else:
                xCheck, yCheck = self.checkpoints[self.step-1][0]*40, self.checkpoints[self.step-1][1]*40
            if (xCheck-self.speed <= self.x <= xCheck+self.speed and
                yCheck-self.speed <= self.y <= yCheck+self.speed):
                self.x, self.y = xCheck, yCheck
                self.step -= 1
                if self.step == -1:
                    self.dead = True
                    print("Woah, slow down buddy\nThat's too many Smoke Bombs for 1 man to handle")
            else:
                #Else move it towards it's next checkpoint
                if self.x < xCheck:
                    self.x += int(self.speed*speed*self.speedModifier[0])
                elif self.x > xCheck:
                    self.x -= int(self.speed*speed*self.speedModifier[0])
                elif self.y > yCheck:
                    self.y -= int(self.speed*speed*self.speedModifier[0])
                elif self.y < yCheck:
                    self.y += int(self.speed*speed*self.speedModifier[0])

        if self.fire[0] == True:
            self.fire[1] -= 1
            self.fire[3] -= 1
            if self.fire[1] <= 0:
                self.fire[1] = self.fire[2] + 0
                if self.rank >= 6 and self.rank - self.fire[4] < 6:
                    self.duplicate = True
                self.ReRank(self.rank - self.fire[4])
                

            if self.fire[3] <= 0:
                self.fire[0] = False

        if self.speedModifier[1] <= 0:  
            if not self.permaslow:
                self.speedModifier[0] = 1
            else:
                self.speedModifier[0] = 0.75
        else:
            self.speedModifier[1] -= 1

        self.confused = False

        return Lives

    def update(self, Lives, speed, gameDisplay):
        self.draw(gameDisplay)
        if speed != 0:
            if self.cooldown <= 0:
                Lives = self.movement(Lives, speed)
            else:
                self.cooldown -= 1

        #Updating hitbox
        self.rect.top = self.y
        self.rect.bottom = self.y + self.height
        self.rect.left = self.x
        self.rect.right = self.x + self.width

        
        return Lives


mainMenu.py
import sys, main, os
 
import pygame
from pygame.locals import *
 
pygame.init()
 
fps = 60
fpsClock = pygame.time.Clock()
 
width, height = 840, 580
gameDisplay = pygame.display.set_mode((width, height))
font_20 = pygame.font.Font("IBMPlexSans-Regular.ttf", 20)
font_30 = pygame.font.Font("IBMPlexSans-Regular.ttf", 30)
font_60 = pygame.font.Font("IBMPlexSans-Regular.ttf", 60)

def load_images(path_to_directory):
    """Loads all images"""
    images = {}
    for dirpath, dirnames, filenames in os.walk(path_to_directory):
        for name in filenames:
            if name.endswith('.png'):
                key = name[:-4]
                img = pygame.image.load(os.path.join(dirpath, name)).convert_alpha()
                images[key] = img
    return images
 
def MainMenu():

    Images = load_images("Images")
    game_run = True
    screen = "Main"
    step = 0
    cooldown = False
    Names = [["Classic", "Twisty Towers"]]
    SaveNames = ["Default", "Twisty"]
    Diff = ["Easy", "Medium", "Hard"]
    file = open("SaveFile/saveFile.txt", "r")
    data = file.readline().split("#")
    if len(data) != 4:
        data = ["0", "Medium", "0", "Medium"]
    
    while game_run:
        gameDisplay.fill((0, 150, 0))
        pos, pressed = pygame.mouse.get_pos(), pygame.mouse.get_pressed()

        for event in pygame.event.get():
            if event.type == QUIT:
              pygame.quit()
              sys.exit()

        if screen == "Main":
            gameDisplay.blit(font_60.render("Block Tower Defense", True, (0, 0, 0)), (130, 150))

            pygame.draw.rect(gameDisplay, (150, 0, 0), (340, 320, 150, 75), 0)
            pygame.draw.rect(gameDisplay, (100, 0, 0), (340, 320, 150, 75), 2)
            gameDisplay.blit(font_20.render("Play", True, (0, 0, 0)), (395, 340))
            if 340 <= pos[0] <= 490 and 320 <= pos[1] <= 395 and pressed[0] == 1:
                screen = "Map Select"
                cooldown = True


        elif screen == "Map Select":

            if step == 0:
                gameDisplay.blit(pygame.transform.scale(Images["Meadows"], (200, 150)), (100, 100))
                gameDisplay.blit(pygame.transform.scale(Images["TwistyTowers"], (200, 150)), (350, 100))
                if 100 <= pos[0] <= 300 and 100 <= pos[1] <= 250 and pressed[0] == 1 and not cooldown:
                    screen = "Dif Select"
                    Map = "Classic"
                    MapName = "Default"
                if 350 <= pos[0] <= 550 and 100 <= pos[1] <= 250 and pressed[0] == 1 and not cooldown:
                    screen = "Dif Select"
                    Map = "Twisty Towers"
                    MapName = "Twisty"
                    
 
            for i in range(2):
                pygame.draw.rect(gameDisplay, (100, 100, 100), (100+250*i, 100, 200, 150), 2)
                gameDisplay.blit(font_20.render(Names[step][i], True, (0, 0, 0)), (200+250*i-int(font_20.size(Names[step][i])[0]/2), 70))
                
            if pressed[0] == 0:
                cooldown = False


            pygame.draw.rect(gameDisplay, (100, 0, 0), (10, 10, 100, 75), 0)
            pygame.draw.rect(gameDisplay, (200, 0, 0), (10, 10, 100, 75), 3)
            gameDisplay.blit(font_30.render("Back", True, (0, 0, 0)), (25, 20))
            if 10 <= pos[0] <= 110 and 10 <= pos[1] <= 85 and pressed[0] == 1 and not cooldown:
                screen = "Main"
                cooldown = True
        elif screen == "Dif Select":
            gameDisplay.blit(font_60.render(Map, True, (0, 0, 0)), (400-int(font_60.size(Map)[0]/2), 20))


            #Different Difficulty Buttons
            for i in range(3):
                pygame.draw.rect(gameDisplay, (0, 100, 0), (50+250*i, 400, 200, 150), 0)
                pygame.draw.rect(gameDisplay, (0, 200, 0), (50+250*i, 400, 200, 150), 3)
                gameDisplay.blit(font_30.render(Diff[i], True, (0, 0, 0)), (150+250*i-int(font_30.size(Diff[i])[0]/2), 440))

                if 50+250*i <= pos[0] <= 50+250*i+200 and 400 <= pos[1] <= 550 and pressed[0] == 1 and not cooldown:
                    main.game_loop(False, MapName, Diff[i])

            #The Continue Button
            pygame.draw.rect(gameDisplay, (0, 100, 0), (50, 290, 120, 95), 0)
            pygame.draw.rect(gameDisplay, (0, 200, 0), (50, 290, 120, 95), 3)
            gameDisplay.blit(font_20.render("Continue", True, (0, 0, 0)), (60, 330))


            #Displays wave and difficulty of your continue
            gameDisplay.blit(font_20.render("Wave: " + str(data[SaveNames.index(MapName)*2]), True, (0, 0, 0)), (60, 310))
            if data[SaveNames.index(MapName)*2+1] != "0":
                gameDisplay.blit(font_20.render(str(data[SaveNames.index(MapName)*2+1]), True, (0, 0, 0)), (60, 290))
            else:
                gameDisplay.blit(font_20.render("Medium", True, (0, 0, 0)), (60, 290))

            if 50 <= pos[0] <= 170 and 290 <= pos[1] <= 385 and pressed[0] == 1 and not cooldown:
                main.game_loop(True, MapName)
            
            #The Back Button
            pygame.draw.rect(gameDisplay, (100, 0, 0), (10, 10, 100, 75), 0)
            pygame.draw.rect(gameDisplay, (200, 0, 0), (10, 10, 100, 75), 3)
            gameDisplay.blit(font_30.render("Back", True, (0, 0, 0)), (25, 20))

            if 10 <= pos[0] <= 110 and 10 <= pos[1] <= 85 and pressed[0] == 1 and not cooldown:
                screen = "Map Select"
                cooldown = True

            if pressed[0] == 0:
                cooldown = False

        pygame.display.flip()
        fpsClock.tick(fps)

if __name__ == "__main__":
    MainMenu()
